cnf-certification-test/accesscontrol/rbac/roles.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (C) 2022 Red Hat, Inc.
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation; either version 2 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+package rbac
+
+import (
+	"strings"
+
+	"github.com/test-network-function/cnf-certification-test/pkg/stringhelper"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+type RoleRule struct {
+	Resource RoleResource
+	Verb     string
+}
+type RoleResource struct {
+	Group, Name string
+}
+
+type CrdResource struct {
+	Group, SingularName, PluralName string
+	ShortNames                      []string
+}
+
+func GetCrdResources(crds []*apiextv1.CustomResourceDefinition) (resourceList []CrdResource) {
+	for _, crd := range crds {
+		var aResource CrdResource
+		aResource.Group = crd.Spec.Group
+		aResource.SingularName = crd.Spec.Names.Singular
+		aResource.PluralName = crd.Spec.Names.Plural
+		aResource.ShortNames = crd.Spec.Names.ShortNames
+		resourceList = append(resourceList, aResource)
+	}
+	return resourceList
+}
+
+func GetAllRules(aRole *rbacv1.Role) (ruleList []RoleRule) {
+	for _, aRule := range aRole.Rules {
+		for _, aGroup := range aRule.APIGroups {
+			for _, aResource := range aRule.Resources {
+				for _, aVerb := range aRule.Verbs {
+					var aRoleRule RoleRule
+					aRoleRule.Resource.Group = aGroup
+					aRoleRule.Resource.Name = aResource
+					aRoleRule.Verb = aVerb
+					ruleList = append(ruleList, aRoleRule)
+				}
+			}
+		}
+	}
+	return ruleList
+}
+
+// Checks the resource name in the role against plural, singular and short names
+func equalResource(aResource CrdResource, aRule RoleRule) bool {
+	// remove subresources to keep only resource name
+	ruleBaseResourceName := strings.Split(aRule.Resource.Name, "/")[0]
+	return aResource.Group == aRule.Resource.Group && (aResource.SingularName == ruleBaseResourceName ||
```

Comment:
I think most of these comparisons are not needed. I don't think you can have different CRDs with the same pluralName and ApiGroup but with different SingularName. Same with the short names. If I'm not wrong, this implementation should be good enough:
```
func isResourceInRoleRule(crd CrdResource, roleRule RoleRule) bool {
	// remove subresources to keep only resource (plural) name
	ruleResourcePluralName := strings.Split(roleRule.Resource.Name, "/")[0]

	return crd.Group == roleRule.Resource.Group && crd.PluralName == ruleResourcePluralName
}
```

Reply:
This code is comparing the following between a CRD and a rule (inside a role):
- rule's resource group with CRD group
- rule's resource name with either CRD plural (more usual) or Singular or short (any in the list ) names 
For instance, if comparing only the plural form and the role's rule is using the short name this function will not match but it should.

Reply:
Ok, I just thought that as per k8s rbac docs [referring to resources](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources) section, every item in the resources list in role's rules should use the plural form always. It can also be seen doing:
```
oc get roles.rbac.authorization.k8s.io -A -o json | jq -r '.items[] | .rules[] | .resources[]
```

Reply:
I think you are right, just did a test with a new user and creating a single restrictive rule to allow pod listing. When using the singular or the short resource names, the user was not able to list pods. only the plural resource worked. Looks like singular and short are used only by kubectl. I will use you suggestion